cfgmgr/vlanmgr.cpp

```diff
@@ -32,144 +32,162 @@ VlanMgr::VlanMgr(DBConnector *cfgDb, DBConnector *appDb, DBConnector *stateDb, c
     SWSS_LOG_ENTER();
 
     // Initialize Linux dot1q bridge and enable vlan filtering
-    stringstream cmd;
-    string res;
+    // The command should be generated as:
+    // /bin/bash -c "/sbin/ip link del Bridge 2>/dev/null ;
+    //               /sbin/ip link add Bridge up type bridge &&
+    //               echo 1 > /sys/class/net/Bridge/bridge/vlan_filtering &&
+    //               /sbin/bridge vlan del vid 1 dev Bridge self"
 
-    cmd << IP_CMD << " link del " << DOT1Q_BRIDGE_NAME;
-    swss::exec(cmd.str(), res);
-
-    cmd.str("");
-    cmd << IP_CMD << " link add " << DOT1Q_BRIDGE_NAME << " up type bridge";
-    EXEC_WITH_ERROR_THROW(cmd.str(), res);
-
-    cmd.str("");
-    cmd << ECHO_CMD << " 1 > /sys/class/net/" << DOT1Q_BRIDGE_NAME << "/bridge/vlan_filtering";
-    int ret = swss::exec(cmd.str(), res);
     /* echo will fail in virtual switch since /sys directory is read-only.
      * need to use ip command to setup the vlan_filtering which is not available in debian 8.
-     * Once we move sonic to debian 9, we can use IP command by default */
-    if (ret != 0)
-    {
-        cmd.str("");
-        cmd << IP_CMD << " link set " << DOT1Q_BRIDGE_NAME << " type bridge vlan_filtering 1";
-        EXEC_WITH_ERROR_THROW(cmd.str(), res);
-    }
-
-    cmd.str("");
-    cmd << BRIDGE_CMD << " vlan del vid " << DEFAULT_VLAN_ID << " dev " << DOT1Q_BRIDGE_NAME << " self";
-    EXEC_WITH_ERROR_THROW(cmd.str(), res);
+     * Once we move sonic to debian 9, we can use IP command by default
+     * ip command to create bridge with filtering:
+     * /sbin/ip link add Bridge up type bridge vlan_filtering 1
+     */
+    const std::string cmds = std::string("")
+      + BASH_CMD + " -c \""
+      + IP_CMD + " link del " + DOT1Q_BRIDGE_NAME + " 2>/dev/null; "
+      + IP_CMD + " link add " + DOT1Q_BRIDGE_NAME + " up type bridge && "
+      + "echo 1 > /sys/class/net/" + DOT1Q_BRIDGE_NAME + "/bridge/vlan_filtering && "
```

Comment:
as explained in the comment, echo will fail in virtual switch since /sys directory is read-only. therefore, you need keep the ip link as the fallback, because in virtual switch you are using the kernel of host, that can support vlan_filtering command.

Reply:
Debian 8 doesn't have support of ip link set Bridge vlan_filtering command. The ip route 2 sources doesn't have anything related to vlan_filtering.
Debian 9 contains updated iproute2 which allows set vlan_filtering from command line without echoing to /sys.
So I left a comment
Once we move sonic to debian 9, we can use IP command by default
ip command to create bridge with filtering:                     
``` /sbin/ip link add Bridge up type bridge vlan_filtering 1```

Reply:
as I said, "echo 1 > ...bridge/vlan_filtering" does not work for virtual switch docker as the /sys is read only. The only way to make it work is to use ip link command. In virtual switch docker, I installed the latest iproute2 command so it works. remove the fallback will break virtual switch.

Reply:
I specifically added this fallback method for virtual switch. 

https://github.com/Azure/sonic-swss/commit/57a98acb893c73063792bf6d7adf8bca71f8b9d6

Reply:
I'm not familiar with virtual switch but I'm workarounding by executing `$mount -o remount,rw /sys` to make sysfs writable in privileged docker container.

Maybe you could do the same to remove this fallback method?

Reply:
the fallback logic will work for debian 9 without code change. it is better to keep it. i tried many methods in virtual switch env, even the one you suggested, none of them works.

Reply:
oh, I got it. sorry for the noise.